storage/multicdnstorage.py

```diff
@@ -0,0 +1,195 @@
+import logging
+from typing import Dict
+
+from storage.basestorage import BaseStorageV2, InvalidConfigurationException
+from util.ipresolver import GEOIP_CONTINENT_CODES
+
+logger = logging.getLogger(__name__)
+
+from storage.storagecontext import StorageContext
+from storage.storagedriverclasses import STORAGE_DRIVER_CLASSES
+
+VALID_RULE_KEYS = ["namespace", "continent", "target"]
+
+
+class MultiCDNStorage(BaseStorageV2):
```

Comment:
Could you subclass S3Storage instead and create an interface for your CDNs (abstract `get_direct_download_url` method)?
That would save you from redefining a bunch of method that's not relevant to the CDN.
Then you would just need to initialize your providers which implement that interface.

Reply:
It'll be also easier to write tests if you interface it. That way we can just test the routing logic, instead of having to deal directly with the implementations for CloudFlare and CloudFront.

Reply:
I don't want to subclass `S3Storage` because that would mean I'll have to add things like bucket, access and secret key in the multiCDN provider config. I want all of that to be transparent to the MultiCDN provider as it's just a proxy and everything else should be handled by the underlying sub-providers.

In this case, I'm using the default provider for other operations but in the future that could just be another config value.

Reply:
You still need S3 for other storage ops other than blob downloads though. How are you configuring your storage from the snippet in the description?

If it's proxy then it shouldn't be subclassing BaseStorage IMO, since it's not really a storage class. It should have its own separate abstract class/interface, which the existing storage implementations would reference.